api/encoding/call_option.go

```diff
@@ -26,32 +26,84 @@ package encoding
 // Encoding authors should accept yarpc.CallOptions and convert them to
 // encoding.CallOptions to use with NewOutboundCall. This will keep the
 // API for service authors simple.
-type CallOption struct{ apply func(*OutboundCall) }
+type CallOption struct {
+	opt callOption
+}
+
+type callOption interface {
+	apply(*options)
+}
+
+type options struct {
+	headers         []keyValuePair
+	shardKey        *string
+	routingKey      *string
+	routingDelegate *string
+	responseHeaders *map[string]string
+}
+
+type responseHeadersOptions struct {
+	h *map[string]string
+}
+
+func (r responseHeadersOptions) apply(opts *options) {
+	opts.responseHeaders = r.h
+}
 
 // ResponseHeaders specifies that headers received in response to this request
 // should replace the given map.
 func ResponseHeaders(h *map[string]string) CallOption {
-	return CallOption{func(o *OutboundCall) { o.responseHeaders = h }}
+	return CallOption{responseHeadersOptions{h}}
+}
+
+type headerOption struct {
+	kv keyValuePair
+}
+
+func (r headerOption) apply(opts *options) {
+	opts.headers = append(opts.headers, r.kv)
 }
 
 // WithHeader adds a new header to the request.
 func WithHeader(k, v string) CallOption {
-	return CallOption{func(o *OutboundCall) {
-		o.headers = append(o.headers, keyValuePair{k: k, v: v})
-	}}
+	return CallOption{headerOption{kv: keyValuePair{k: k, v: v}}}
```

Comment:
based the [guide](https://github.com/uber-go/guide/blob/e4602992ffbc465ea14bce5501cdbfc361fd8c96/style.md#functional-options), those WithXXX func should return the interface type `callOption`?

Reply:
We can not do this since it is a public API.